### ISSUE TITLE
pytorch_scatter improvement

### ISSUE THREAD

**Issue by tchaton** [getkeops/keops#26]
Hello there,

I would like to ask some questions about keops.

I have been using a lot pytorch geometric for my work on graphs.
It uses pytorch scatter has its core: https://github.com/rusty1s/pytorch_scatter
And the MessagePassing https://github.com/rusty1s/pytorch_geometric/blob/master/torch_geometric/nn/conv/message_passing.py which is using torch.select_index for the message.

I also found this paper implementing a smarter hierarchical scatter method 
![image](https://user-images.githubusercontent.com/12861981/65950315-b7d37900-e435-11e9-90ee-d0c9530f2883.png)

I was wondering if keops could be used to implement a symbolic message function and maybe also the HAG aggregation within a new pytorch_scatter.

I would not only reduce drastically the memory, but also could speed up training / inference.

What are you thoughts on that ?

Best,
Thomas Chaton.

**Comment by tchaton** [comment]
``````
# Create two arrays with 3 columns and a (huge) number of lines, on the GPU
import torch
x = torch.randn(1000000, 3, requires_grad=True).cuda()
y = torch.randn(2000000, 3).cuda()

# Turn our Tensors into KeOps symbolic variables:
from pykeops.torch import LazyTensor
x_i = LazyTensor( x[:,None,:] )  # x_i.shape = (1e6, 1, 3)
y_j = LazyTensor( y[None,:,:] )  # y_j.shape = ( 1, 2e6,3)

# We can now perform large-scale computations, without memory overflows:
D_ij = ((x_i - y_j)**2).sum(dim=2)  # Symbolic (1e6,2e6,1) matrix of squared distances
K_ij = (- D_ij).exp()               # Symbolic (1e6,2e6,1) Gaussian kernel matrix
# Note that in fact nothing has been computed yet, everything will be done in the final reduction step

# Now we come back to vanilla PyTorch Tensors or NumPy arrays using
# reduction operations such as .sum(), .logsumexp() or .argmin().
# Here, the kernel density estimation   a_i = sum_j exp(-|x_i-y_j|^2)
# is computed using a CUDA online map-reduce routine that has a linear
# memory footprint and outperforms standard PyTorch implementations
# by two orders of magnitude. All actual computations are performed at this step.
a_i = K_ij.sum(dim=1)  # Genuine torch.cuda.FloatTensor, a_i.shape = (1e6, 1),
g_x = torch.autograd.grad((a_i ** 2).sum(), [x])  # KeOps supports autograd!
``````

As you said there, do you think the sparse scatter could be this reduction operations ?

**Comment by jeanfeydy** [comment]
Hi @tchaton,

Thanks a lot for the links and references!
Two quick thoughts:
1. Coincidence, I just implemented a `scattered_sum` reduction for issue #24 (still have to implement the gradient and some doc + unit tests before merging in master). This thread is a good illustration of the way the current framework can be extended to support sparse-like problems. If you'd like to make a similar feature requests, with specific "toy examples" to optimize, please feel free to do so. Note however that the KeOps schemes will always be suited to full and block-sparse schemes, with loads of interactions to compute and nearly-contiguous memory accesses. I don't think we'll ever be able to be competitive on "Twitter-like" graph structures, with super-sparse and random connectivities.

2. Our short-term priority has been to diffuse KeOps in the Kernel and Gaussian Process communities: we're mathematicians by trade, and their problems are super close to what we know as we share a common language, etc. That's why we've put a lot of effort into kernel-centric tutorials, R backends and JMLR/AiStats-like publications. However, I just got hired as a PostDoc for three years in Michael Bronstein's team (at Imperial College, we're I'm sitting right now), where all the students use PyTorch_Geometric: in the next few months/years, extending KeOps in a direction that is suited to Geometric/Graph deep learning will be my top priority.

Note that I will be pretty busy with paper writing / redaction until ~December: I'll be happy to read what you and @rusty1s can send to me on the subject, but probably won't be able to implement anything seriously before January :-)

Best regards,
Jean

**Comment by tchaton** [comment]
@jeanfeydy  I have sent you an email jean.feydy@gmail.com. I hope it finds you well.
